### DISCUSSION TITLE
Parent/child result set crash, getters are off by one!

### DISCUSSION THREAD

**Discussion by amyworrall** [ricomariani/CG-SQL-author#73] [Q&A] [answered]
I'm not sure if this is something I'm doing wrong, or if it's a bug. Probably the former, but I'm a bit stumped!

I'm trying to build parent/child result set support. Regarding my environment, I'm on iOS, i'm using `cqlrt_cf`, and I've built a few layers of my own codegen on top of the JSON output (to provide an easy Swift interface).

For doing the CQL compilation, I'm using the following as part of my build process (before moving on to my own codegen):

```
cc -E -x c -Wno-invalid-pp-token $INPUT_FILE_NAME | cql  --cqlrt cqlrt_cf.h --cg "$DERIVED_SOURCES_DIR/$INPUT_FILE_BASE.cql.h" "$DERIVED_SOURCES_DIR/$INPUT_FILE_BASE.cql.c"

cc -E -x c -Wno-invalid-pp-token $INPUT_FILE_NAME | cql --rt json_schema --cqlrt cqlrt_cf.h --cg "$DERIVED_SOURCES_DIR/$INPUT_FILE_BASE.json"

cc -E -x c -Wno-invalid-pp-token $INPUT_FILE_NAME | cql --rt objc_mit --objc_c_include_path $INPUT_FILE_BASE.cql.h --cqlrt cqlrt_cf.h --cg "$DERIVED_SOURCES_DIR/$INPUT_FILE_BASE.objc.h"
```

I'm trying to build parent/child result sets. I've had things working using the cursor-based approach, but I'm having trouble with the out union based approach.

This works:

```
create proc TestParent()
begin
  SELECT roomID, name FROM rooms ORDER BY name;
end;

create proc TestChild(id UUID NOT NULL)
begin
  SELECT * FROM tasks WHERE roomID == id;
end;

create proc TestParentChild()
begin
  -- the result is like the parent with an extra column for the child
  declare result cursor like (like TestParent, child object<TestChild set>);

  -- call the parent and loop over the results
  declare P cursor for call TestParent();
  loop fetch P
  begin
     -- compute the child for each P and then emit it
     fetch result from values(from P, TestChild(P.roomID));
     out union result;
  end;
end;
```

If I do the above, I can get a child resultset from the child field of a row of my TestParentChild query, and access all of its rows/columns.

However, this does not work:

```
create proc TestParent()
begin
  SELECT roomID, name FROM rooms ORDER BY name;
end;

create proc TestChild()
begin
  SELECT roomID, tasks.taskID as thisIsATask FROM tasks;
end;

create proc TestParentChild()
begin
  out union call TestParent() join call TestChild() using (roomID) as tasks;
end;
```

If I do the latter, I get an `EXC_BAD_ACCESS` when I try to make use of the nested result set. More specifically, when I call `TestChild_get_roomID(cResultSet, row)` I get an `EXC_BAD_ACCESS`.

However, if I call `TestChild_get_thisIsATask(cResultSet, row)`, I get the value I should have got for get_roomID! I've tested adding a few more columns to the output of TestChild, and each accesor always returns the value of one column to the left of what it should be, with the first one always crashing.

Any ideas where I should look to work this out?

**Comment by ricomariani** [comment]
Unfortunately I don't have a Mac anymore so I can't run cqlrt_cf :(

But let's dig into the codegen.  The code is going to rely on the C wrappers because that's how we roll.

This tells us that the main result set from `TestParentChild` has 2 references in it and they start at `name`.  That's correct.  There are 3 fields total.

```C
#define TestParentChild_refs_offset cql_offsetof(TestParentChild_row, name) // count = 2

static cql_uint16 TestParentChild_col_offsets[] = { 3,
  cql_offsetof(TestParentChild_row, roomID),
  cql_offsetof(TestParentChild_row, name),
  cql_offsetof(TestParentChild_row, tasks)
};
```

The data types for the reader are correct also:

```C
uint8_t TestParentChild_data_types[TestParentChild_data_types_count] = {
  CQL_DATA_TYPE_INT32 | CQL_DATA_TYPE_NOT_NULL, // roomID
  CQL_DATA_TYPE_STRING, // name
  CQL_DATA_TYPE_OBJECT | CQL_DATA_TYPE_NOT_NULL, // tasks
};
```

The row structure itself also seems correct

```C
typedef struct TestParentChild_row {
  cql_bool _has_row_;
  cql_uint16 _refs_count_;
  cql_uint16 _refs_offset_;
  cql_int32 roomID;
  cql_string_ref _Nullable name;
  cql_object_ref _Nonnull tasks;
} TestParentChild_row;
```

Next we look at the getters, each of these looks like it's pulling out the right field also.

```C
cql_int32 TestParentChild_get_roomID(TestParentChild_result_set_ref _Nonnull result_set, cql_int32 row) {
  TestParentChild_row *data = (TestParentChild_row *)cql_result_set_get_data((cql_result_set_ref)result_set);
  return data[row].roomID;
}

cql_string_ref _Nullable TestParentChild_get_name(TestParentChild_result_set_ref _Nonnull result_set, cql_int32 row) {
  TestParentChild_row *data = (TestParentChild_row *)cql_result_set_get_data((cql_result_set_ref)result_set);
  return data[row].name;
}

TestChild_result_set_ref _Nonnull TestParentChild_get_tasks(TestParentChild_result_set_ref _Nonnull result_set, cql_int32 row) {
  TestParentChild_row *data = (TestParentChild_row *)cql_result_set_get_data((cql_result_set_ref)result_set);
  return (TestChild_result_set_ref _Nonnull )data[row].tasks;
}
```

Now the child rowset is simpler even:

```
typedef struct TestChild_row {
  cql_int32 roomID;
  cql_int32 thisIsATask;
} TestChild_row;

cql_int32 TestChild_get_roomID(TestChild_result_set_ref _Nonnull result_set, cql_int32 row) {
  TestChild_row *data = (TestChild_row *)cql_result_set_get_data((cql_result_set_ref)result_set);
  return data[row].roomID;
}

cql_int32 TestChild_get_thisIsATask(TestChild_result_set_ref _Nonnull result_set, cql_int32 row) {
  TestChild_row *data = (TestChild_row *)cql_result_set_get_data((cql_result_set_ref)result_set);
  return data[row].thisIsATask;
}

uint8_t TestChild_data_types[TestChild_data_types_count] = {
  CQL_DATA_TYPE_INT32 | CQL_DATA_TYPE_NOT_NULL, // roomID
  CQL_DATA_TYPE_INT32 | CQL_DATA_TYPE_NOT_NULL, // thisIsATask
};
```

That all looks right.

If I had to guess what's going on here it's that the child getter is getting the parent rowset as an arg?  I can't see your code to verify this is the case though.  You can put a breakpoint on TestChild_get_roomID and verify that the incoming pointer is correct and the row is correct.  But it would assert if the row was out of bounds so I'm guessing the pointer?

It doesn't look like you're using the OBJC wrapper header at all, which is fine, you don't need to.

**Comment by ricomariani** [comment]
Ok I got it.  I see what's going wrong.  This all comes from being lazy.  The test cases for this use CQL to read the child rowset and so they are immune to this problem.

TestChild is supposed to produce this code:

```C
typedef struct TestChild_row {
  cql_int32 roomID;
  cql_int32 thisIsATask;
} TestChild_row;
```

But the cursor it's loaded with has the extra cursor state fields in it:

```
typedef struct TestParentChild___child_cursor__0_row {
  cql_bool _has_row_;
  cql_uint16 _refs_count_;
  cql_uint16 _refs_offset_;
  cql_int32 roomID;
  cql_int32 thisIsATask;
} TestParentChild___child_cursor__0_row;
```

That means the row has extra fields in it!

**Comment by ricomariani** [comment, marked as answer]
You can workaround this problem by adding ` --generate_type_getters` to the cql command line which generates alternate code to read fields.  This is the pattern that is used inside Meta because it is more JNI friendly and it is also immune to this problem.

**Reply by ricomariani** [reply]
Note that you no longer need to do this. In fact you can't.  --generate_type_getters is now the default/only option so there's no switch for it anymore.  The other option had simpler codegen but didn't always work.

**Comment by ricomariani** [comment]
repro:

```sql
create proc init()
begin
  create table tasks(
   taskID int!,
   roomID int!
  );

  create table rooms (
    roomID int!,
    name text
  );

  insert into rooms values (1, "foo"), (2, "bar");
  insert into tasks values (100,1), (101,1), (200,2);
end;

[[private]]
create proc TestParent()
begin
  SELECT roomID, name FROM rooms ORDER BY name;
end;

create proc TestChild()
begin
  SELECT roomID, tasks.taskID as thisIsATask FROM tasks;
end;

create proc TestParentChild()
begin
  out union call TestParent() join call TestChild() using (roomID) as tasks;
end;
```

```C
// main.c  (note that this code leaks memory because left out the releases)
#include "cqlrt.h"
#include "x.h"

int main(int argc, char **argv) {
  sqlite3 *db = NULL;
  cql_code rc;

  rc = sqlite3_open(":memory:", &db);
  if (rc) exit(1);

  rc = init(db);
  if (rc) exit(2);

  TestParentChild_result_set_ref parent;

  rc = TestParentChild_fetch_results(db, &parent);
  if (rc) exit(3);

  cql_int32 count = TestParentChild_result_count(parent);
  printf("found %d rows\n", count);

  for (int i = 0; i < count; i++) {
    cql_int32 roomID = TestParentChild_get_roomID(parent, i);
     cql_string_ref name_ref = TestParentChild_get_name(parent, i);

     cql_alloc_cstr(cstr, name_ref);
       printf("roomID: %d name: %s\n",  roomID, cstr);
     cql_free_cstr(cstr, name_ref);

     TestChild_result_set_ref child = TestParentChild_get_tasks(parent, i);
     cql_int32 children = TestChild_result_count(child);

     for (int j = 0; j < children; j++) {
       cql_int32 child_room = TestChild_get_roomID(child, j);
       cql_int32 child_task = TestChild_get_thisIsATask(child, j);
       printf("  room:%d task:%d\n", child_room, child_task);
     }
  }

  return 0;
}
```

build:

```bash
out/cql --in x.sql --cg x.h x.c --generate_type_getters
cc -g -o x main.c x.c cqlrt.c -lsqlite3
```

Correct output:

```
found 2 rows
roomID: 2 name: bar
  room:2 task:200
roomID: 1 name: foo
  room:1 task:100
  room:1 task:101
```

**Comment by ricomariani** [comment]
The way to fix this is to make better shaped results with `out union` generally which is actually goodness.  It's just annoying so I never did it...

**Comment by ricomariani** [comment]
This is a righteous bug report, thank you!

**Comment by amyworrall** [comment]
Hi! Many thanks for all your investigation. I can confirm that the workaround with `--generate_type_getters` fixes things for me!

**Comment by amyworrall** [comment]
Are there any downsides to generate_type_getters? Just curious as to why CGSQL still has both…

**Comment by ricomariani** [comment]
They are slightly bigger.

________________________________

**Comment by ricomariani** [comment]
I've been thinking about how best to fix this since yesterday and I'm coming to the conclusion that the best choice is to make type getters the default/only option.  The other option has an ABI that is too strict and even CQL itself has a hard time honoring it.  It's not a coincidence that type getters is the only option we ever used in Meta.  Even though it was the 2nd option in this case "2nd" means "about 2 weeks after the first".

This would let me delete a bunch of code and make maintenance simpler at the cost of a little runtime perf but saving memory.  The other option is to always make rowsets the fattest option whether we need the fat or not.  I am not liking that at all.  Plus they have to stay identical and if new things we want to put in rowsets came along we could be stuck.

I'll probably just do the delete this weekend.

**Comment by amyworrall** [comment]
Sounds good to me. I'm targetting iPhones -- they've got plenty of perf :)

**Comment by ricomariani** [comment]
https://github.com/ricomariani/CG-SQL-author/pull/74

The original target for CQL was iOS.  For Messenger on iOS in particular.  cqlrt_cf is a clean version of the original runtime also.  The first version was for iOS.

**Comment by ricomariani** [comment]
Note: when the PR lands --generate_type_getters will be no longer be a valid option, I removed it, it's now the only option.  So you can just get rid of it in your build.

**Reply by ricomariani** [reply]
FWIW I also noticed that you were using cc -E etc. to preprocess the CQL.  With `@include` and `@macro` you don't have to do that anymore.  You might want to switch to that and simplify your build further.

**Comment by amyworrall** [comment]
> The original target for CQL was iOS. For Messenger on iOS in particular. cqlrt_cf is a clean version of the original runtime also. The first version was for iOS.

Not sure if you recognise my name, but I'm one of the ones who led Work Chat development at Meta (and was a thorn in your side about schemas!). Am still at Meta but on a different team now. (My use of CQL here is for a personal project, not work stuff!)

**Reply by ricomariani** [reply]
Be fun to catch up!  I'm curious what you're up to with it.  I was wondering how you even found it :D

**Comment by ricomariani** [comment]
Tests added for this case: https://github.com/ricomariani/CG-SQL-author/commit/8bb0f6a136498279e15a7270e04767584ea5b6e6
